### PR TITLE
Add preview for subject with text location

### LINKED ISSUE #4936: Subject preview not showing anything for text location
## Package
- app-project (maybe lib-react-components)

## Describe the bug
For subjects with a first location of type text the subject preview doesn't show anything.
Related [Talk post](https://www.zooniverse.org/talk/14/2983249?comment=4900223).

Live ⚠️ production ⚠️ project for reference - https://www.zooniverse.org/projects/md68135/notes-from-nature-labs

## To Reproduce
1. Go to https://www.zooniverse.org/projects/md68135/notes-from-nature-labs
2. Note the recents section shows the subjects with IDs but blank media

Also applies on classify page "Your recent classifications" section

app-project [SubjectPreview](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/app-project/src/shared/components/SubjectPreview/SubjectPreview.js) component
lib-classifier [Media](https://github.com/zooniverse/front-end-monorepo/tree/master/packages/lib-react-components/src/Media) component

## Expected behavior
The subject preview should show something. If only a text location, maybe the text icon used in frame controls should show. If the subject is an image and text subject, should we preview the first image location instead of a text icon?

## Screenshots
subjectPreviews

### PR DESCRIPTION
**Pull request by mcbouslog**
## Package
- app-project
- lib-react-components

## Linked Issue and/or Talk Post
Closes #4936 

## Describe your changes
- lib-react-components - add Text component within Media component
- lib-react-components - add useTextData hook for Text component
- app-project - refactor SubjectPreview, project home page RecentSubjects, and classify page RecentSubjects for text subjects

## How to Review
_Helpful explanations that will make your reviewer happy:_
- What Zooniverse project should my reviewer use to review UX?
  - https://local.zooniverse.org:3000/projects/markb-panoptes/digileap-testing-staging?env=staging - project home/recents and classify/recents
- What user actions should my reviewer step through to review this PR?
  - review storybooks, comparing to other subject types
  - make a few classifications here - https://frontend.preview.zooniverse.org/projects/markb-panoptes/digileap-testing-staging/classify/workflow/3538?env=staging
  - note existing text subject previews on project home page and classify page
  - note text subject previews locally - https://local.zooniverse.org:3000/projects/markb-panoptes/digileap-testing-staging?env=staging
- Which storybook stories should be reviewed?
  - http://localhost:6007/?path=/story/components-media--text-media
  - http://localhost:9001/?path=/story/project-app-screens-project-home-recent-subjects--text
  - http://localhost:9001/?path=/story/project-app-screens-classify-recent-subjects--text
  - http://localhost:9001/?path=/story/project-app-shared-subject-preview--text
- Are there plans for follow up PR’s to further fix this bug or develop this feature? no

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [ ] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

## Refactoring
- [ ] The PR creator has described the reason for refactoring
- [ ] The refactored component(s) continue to work as expected

### CODE REVIEW

**Comment by shaunanoordin** [comment]
## Quick Review

A quick code read looks good. As far as I can tell, the main code changes are:
- Added: packages/lib-react-components/src/Media/components/Text/Text.js
- Added: packages/lib-react-components/src/hooks/useTextData.js

Both seem pretty straightforward, and I don't expect any issues. I'm running this locally now for testing.

**Review by shaunanoordin — APPROVED** [review]
## PR Review

Packages: `lib-react-components` (main code changes), `app-project` (most visible changes)

This PR allows Text subjects [(example)](https://panoptes-uploads.zooniverse.org/subject_location/f5506d1c-a0e9-4aba-a418-6a6c46a7731a.txt) to be rendered/previewed in the `<Media>` component. This is most visible on the Project's pages "Recent Subjects", and Classify page's functionally similar "Your Recent Classifications"

- Text Subjects are now being correctly previewed/rendered in `<Media>` 👍 
  - Previewed text can be scrolled using mouse scroll. (Move cursor onto text that's too big for the Media component, then scroll up/down to view the full text)
  - ❓ (non-blocking) Quirk: however, there's no way to scroll using _keyboard navigation_ as far as I can tell. (see Dev Notes)
- Text Subjects on the Recent Subjects/Your Recent Classifications lists continue to function as links
- ❔ (non-blocking) Quirk: VoiceOver doesn't actually read text of Subject... which I can't tell is a good thing or a bad thing. 🤷 (see Dev Notes)

Code changes look good, as earlier noted. 

Screenshot: 3x Subjects on the [DigiLeap Testing project's home page's](https://local.zooniverse.org:3000/projects/markb-panoptes/digileap-testing-staging?env=staging) Recent Subjects. Left-most is an Image subject, other two are Text subjects.

Screenshot 2023-11-13 at 21 52 58

Screenshot: 3x Text Subjects on the same project's classify page's Your Recent classifications (logged in as darkeshard)

Screenshot 2023-11-13 at 21 50 41

### Testing

Tested with macOS + Chrome on localhost

Storybook:
- View new Text Subject-related components on app-project and lib-react-component Storybooks.
  - Observe text data is being rendered/previewed as expected

app-project:
- Open [DigiLeap Testing project](https://local.zooniverse.org:3000/projects/markb-panoptes/digileap-testing-staging?env=staging).
  - Basic check: observe text data is being rendered/previewed as expected in Recent Subjects.
  - Mouse functionality: click on a Subject, confirm navigation to Talk page. Use mouse scroll to scroll text.
  - Keyboard functionality: tab-nav to each Subject. Press Enter, confirm navigation. Try to scroll by pressing up/down keys when Subject or Subject's text has focus?
- Open the project's [Classify page ("Highlighter Task Testing" wf)](https://local.zooniverse.org:3000/projects/markb-panoptes/digileap-testing-staging/classify/workflow/3538?env=staging)
  - Login (user darkeshard), then make a few classifications
  - Basic check: observe text data is being rendered/previewed as expected in Your Recent Classifications.
  - Mouse functionality: same as above
  - Keyboard functionality: same as above

### Dev Notes

I've noted a few accessibility-related quirks, but nothing blocking. (i.e. I can't think of better solutions, and any enhancements can be done in a separate PR.)

**Quirk 1:** can't scroll text using keyboard

- I can use Tab to navigate to each Subject, but pressing up/down does not scroll the text content.
- This is because keyboard nav can only focus on the StyledAnchor, whereas the text + overflow CSS is buried much deeper.

```
<a class="StyledAnchor">  <-- keyboard focus only goes up to here
  <section>
    <pre>  <-- overflow: scroll is here
      yabba dabba doo
```

Impact of this quirk: annoying to keyboard users since they don't have access to the same amount of content that mouse users do, but technically this is a Subject _preview_ so technically there's nothing saying we need to see _the whole subject._

Suggested solution: 🤷 

**Quirk 2:** mac VoiceOver doesn't read Subject text

- On Mac, press Cmd+F5 (or whatever your system is set up for) to enable Voice Over
- Tab-nav to any Text Subject in Recent Subjects.
- Note that the VoiceOver reads something like _"link, Subject 12345, SUBJECT 12345"_
  - One might have expected that since this is a _text_ subject, the VoiceOver would read the _text body_ (Subject data), e.g. _"link, Flintstones, meet the Flintstones, they're a modern stone age family, SUBJECT 12345"_
  - As far as I can tell, this is because the containing section has `<section aria-label="subject 12345">`, meaning the aria-label overrides the actual text body.

Impact of this quirk: text subjects are no more visible to VoiceOver users than image subjects.

Possible _(but not necessarily suggested)_ solution: removing the aria-label will cause VoiceOver to read the full Subject text _...but I'm not sure if this is a good thing or not,_ because really long Text Subject will result in a hella long readout. Not sure if users would really want to hear the full Flintstones theme song, or just be happy knowing it's "Subject 12345".

### Status

LGTM 👍

**Comment by mcbouslog** [comment]
@shaunanoordin - really great review notes thank you! 

I've gone back and forth, and tried the initial UI and tried it with giving the `pre` element within the Media Text component tab focus (which allows the text content to be scrolled and announces with screen reader), and I've come to the conclusion I think the UI is preferable with the text content discoverable. I've added tabIndex to the pre element in the Media Text component accordingly. 

I feel like with the pre element tab focus it's consistent with mouse behavior (as you noted), and although it could be a lot for a screen reader to announce, I think it's ok err on giving more info than less. From demos I've watched I get the impression users using a screen reader are comfortable and familiar with skipping content they're not interested in and are very efficient at doing so. However, I'm very open to suggestion otherwise (reverting recent commit and remove pre element tab focus).

**Comment by eatyourgreens** [comment]
VoiceOver isn't reading out the subject text because the link's text content has been overridden with an ARIA label.

File: packages/lib-react-components/src/Media/components/Text/Text.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    content = <p>{error.message}</p>
+  }
+
+  if (data) {
+    content = data
+  }
+
+  return (
+    <StyledBox
+      a11yTitle={alt}
```
Comment (eatyourgreens) [inline packages/lib-react-components/src/Media/components/Text/Text.js:58]: This overrides the text content of the subject, and replaces it with the alt text. 

Alt text isn't needed for text.

**Comment by eatyourgreens** [comment]
You're absolutely right, though, that anything styled with `overflow: scroll` should also be focusable, so that it can be scrolled from the keyboard.

File: packages/lib-react-components/src/Media/components/Text/Text.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      flex={flex}
+      forwardedAs='section'
```
Comment (eatyourgreens) [inline packages/lib-react-components/src/Media/components/Text/Text.js:60]: Is `forwardedAs` still a prop in styled components? For some reason, I thought it had been replaced with `as`.

Comment (mcbouslog) [reply packages/lib-react-components/src/Media/components/Text/Text.js:60]: I think I forget about this every time, but I think if it's just a styled element or just a Grommet element `as` works, but if it's a Grommet component wrapped with styled then [`forwardedAs`](https://styled-components.com/docs/api#forwardedas-prop) I think is necessary.

Comment (eatyourgreens) [reply packages/lib-react-components/src/Media/components/Text/Text.js:60]: That might explain some styling problems I ran into after using `as` in #5521.

File: packages/lib-react-components/src/Media/components/Text/Text.js (hunk 1 above)
Comment (mcbouslog) [reply packages/lib-react-components/src/Media/components/Text/Text.js:58]: I think the a11yTitle is helpful in this instance. Essentially it announces the subject similar to other subject or media previews then with the tabIndex addition if tabbed to announces the text content, so consistent, clear (on first focus, "link subject 123" more clear than "link "This flower is..." (text content)"). Without a11yTitle a screen reader announces "link ..." then announces the text (as to be expected and as you've noted), but I think that would be confusing as a user.

File: packages/lib-react-components/src/Media/components/Text/Text.js (hunk 2 above)
Comment (mcbouslog) [reply packages/lib-react-components/src/Media/components/Text/Text.js:60]: Ahha yes! The StyledAnchors. I think if use `forwardedAs` those should work as expected.

**Pull request merged**